pm/queue.go

```diff
@@ -0,0 +1,148 @@
+type RedeemableEmitter interface {
+	// Redeemable returns a channel that a consumer can use to receive tickets that
+	// should be redeemed
+	Redeemable() chan *SignedTicket
+}
+
+// ticketQueue is a queue of winning tickets that are in line for redemption on-chain.
+// A recipient will have a ticketQueue per sender that it is actively receiving tickets from.
+// If a sender's max float is insufficient to cover the face value of a ticket it is added to the queue.
+// A ticket is pulled from the queue by the recipient when a sender has sufficient max float to cover
+// the next ticket in the queue
+//
+// Based off of: https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/queue.go
+type ticketQueue struct {
+	queue []*SignedTicket
+
+	// queueLen is an internal length counter that keeps track
+	// of the size of the queue. We maintain this counter instead
+	// of reading len(queue) in order to avoid locking the queue loop goroutine
+	queueLen int32
+
+	// in is a channel that receives queued tickets
+	in chan *SignedTicket
+
+	// availableMaxFloat is a channel that the queue uses
+	// to receive the latest available max float value
+	availableMaxFloat chan *big.Int
+
+	// redeemable is a channel that a recipient will receive
+	// redeemable tickets on as a sender's max float becomes
+	// sufficient to cover the face value of tickets
+	redeemable chan *SignedTicket
+
+	quit chan struct{}
+}
+
+func newTicketQueue() *ticketQueue {
+	return &ticketQueue{
+		in:                make(chan *SignedTicket),
+		availableMaxFloat: make(chan *big.Int),
+		redeemable:        make(chan *SignedTicket),
+		quit:              make(chan struct{}),
+	}
+}
+
+// Start initiates the main queue loop goroutine for processing tickets
+func (q *ticketQueue) Start() {
+	go q.startQueueLoop()
+}
+
+// Stop signals the ticketQueue to gracefully shutdown
+func (q *ticketQueue) Stop() {
+	close(q.quit)
+}
+
+// Add adds a ticket to the queue
+func (q *ticketQueue) Add(ticket *SignedTicket) {
+	select {
+	case q.in <- ticket:
+	case <-q.quit:
+		return
+	}
+
+	atomic.AddInt32(&q.queueLen, 1)
+}
+
+// SignalMaxFloat signals to the queue the latest max float for the sender
+// An external caller should call this method whenever a ticket redemption confirms
+// with the most up-to-date max float for the sender
+func (q *ticketQueue) SignalMaxFloat(amount *big.Int) {
+	// If the queue is empty do not signal a new max float
+	// value because there are no tickets to process
+	if q.Length() == 0 {
+		return
+	}
+
+	select {
+	case q.availableMaxFloat <- amount:
+	case <-q.quit:
+		return
+	}
+}
+
+// Redeemable returns a channel that a consumer can use to receive tickets that
+// should be redeemed
+func (q *ticketQueue) Redeemable() chan *SignedTicket {
+	return q.redeemable
+}
+
+// Length returns the current length of the queue
+func (q *ticketQueue) Length() int32 {
+	return atomic.LoadInt32(&q.queueLen)
+}
+
+func (q *ticketQueue) startQueueLoop() {
+	for {
```

Comment:
Isn't this basically a busy loop since nothing blocks? Trying to unpack my understanding here. It looks like there are two things happening within this for-loop:

1. Draining the `in` channel and adding to the queue
2. Checking for max float updates, and if there is one, sending the next ticket (head of queue) to the `redeemable` channel for redemption (`redeemable` is unbuffered; are we confident it will always have a listener?)

The queue may grow if there isn't a max float update at the time the queue is added.

Should these be separated into goroutines running their own blocking selects? This would mean locking around a (function-local) queue. Or maybe even the logic could be combined into one, where we wait for max float updates, then drain `q.in` directly into `q.redeemable` as necessary? Haven't thought about the approach too closely, so I might be off base. Something like this:

```golang
		select {
		case maxFloat := <-q.availableMaxFloat:
			if q.sufficientMaxFloat(maxFloat, nextTicket) {
				select {
				case q.redeemable <- q.in: // `q.in` should be buffered
					continue
				case <-q.quit:
					return
				default:
					continue
				}
			}
		case <-q.quit:
			return
```

BTW if my understanding is inaccurate (or even if it is accurate), it'd probably be helpful to have a brief description in comments of how the queue-loop works, to help folks reviewing the code in the future.

Reply:
> Isn't this basically a busy loop since nothing blocks?

Yeah you're right - this is a busy loop. Previously, the condvar based architecture would have the for loop block until the queue is non-empty and only then would the loop continuously check for max float updates by receiving on `availableMaxFloat`. Now, the loop never blocks - we don't want to block when receiving on `q.in` because this would prevent the loop from receiving on `availableMaxFloat` when the queue is non-empty. We only want to block prior to receiving on `availableMaxFloat` if the queue is empty. I think this is a case where the condvar based architecture feels cleaner (not sure if there is a clean way to exactly replicate the described behavior with channels at the moment).

> The queue may grow if there isn't a max float update at the time the queue is added.

Yep or if winning tickets are received in quick succession such that the sender's max float is not sufficient to cover all of them immediately so they are instead put on the queue and redeemed one by one with the recipient making sure that the sender's max float is sufficient whenever it pops a ticket from the queue.

> redeemable is unbuffered; are we confident it will always have a listener?

The SenderMonitor will [create a queue](https://github.com/livepeer/go-livepeer/pull/926/commits/6bc826f45492a26edae26c61842c3377a875b052#diff-bc4d3ce81466bdbc2bcf8ecbe06525b8R198) for each tracked sender and start a consumer loop that listens on the queue's `redeemable` channel, let's call this `qRedeemable`, for tickets which are then passed into the SenderMonitor's single `redeemable` channel (tickets received from all queues managed by SenderMonitor are fanned into this single channel), let's call this `smRedeemable`, which the Recipient is listening on. The Recipient will redeem each ticket it receives and block until the transaction confirms on-chain - then, the Recipient will receive additional tickets on `smRedeemable`.

I think there is a minor problem here - the send for `qRedeemable` will block if there is a blocking send on `smRedeemable` which will be the case when the Recipient is waiting for a redemption transaction to confirm. This would be fine in a setting where the Recipient is interacting with a single sender. However, in this case, ticket redemptions for *other* senders would slow down the processing of a queue for a particular sender. For example, if the Recipient is waiting for the redemption for sender A and there are tickets waiting to be received on `smRedeemable` for senders B and C, then it takes longer to process the tickets in the queue for senders B and C. An improvement would be for the Recipient not to block while waiting for a redemption to confirm. 

> Should these be separated into goroutines running their own blocking selects? This would mean locking around a (function-local) queue.

If we're locking, I think I'd prefer just sticking with the condvar based architecture.

> Or maybe even the logic could be combined into one, where we wait for max float updates, then drain q.in directly into q.redeemable as necessary?

This was actually what I was thinking of in [this comment](https://github.com/livepeer/go-livepeer/pull/926#discussion_r297364022) describing the use of a buffered channel to maintain the state of the queue instead of relying on a slice. The one thing I don't like about it is that whenever the max float is *not* sufficient to cover the ticket at the head of the queue, we are left with two choices: drop the ticket or pass the ticket back into `q.in`. I don't think we want to drop tickets since its possible that the next max float update is sufficient to cover the ticket at the head of the queue. Passing the ticket back into `q.in` would not only require another select statement that receives on `q.quit` (to exit cleanly if needed), but the send would also need to be in a goroutine because we don't want to block or else the loop won't get to receive on `q.in`. Maintaining the state of the queue in a slice that remains unchanged if the max float is insufficient to cover the ticket at the head of the queue feels simpler.

> BTW if my understanding is inaccurate (or even if it is accurate), it'd probably be helpful to have a brief description in comments of how the queue-loop works, to help folks reviewing the code in the future.

👍

Reply:
> The one thing I don't like about it is that whenever the max float is not sufficient to cover the ticket at the head of the queue, we are left with two choices: drop the ticket or pass the ticket back into q.in

Would checking `sufficientMaxFloat` prior to checking `q.in` address that? As in the example above, `q.in` is nonblocking (via the `default` case), so if empty, we'd continue waiting for the next max float update.

Also, reviewing [recipient.redeemWinningTicket](https://github.com/livepeer/go-livepeer/blob/f46945f0864e52f6ed87c20a1ac668a918a834c7/pm/recipient.go#L323-L333), it does seem that we have the potential to re-queue if the `redeemWinningTicket` function detects the float is insufficient.

Noticed this when I was trying to figure out the "chicken-egg" problem of actually submitting the first ticket from an empty queue, since there'd be no pending transaction waiting to be signaled, so theoretically tickets could be stuck in the queue indefinitely. I see how we're handling that situation now, but it seems like something that could be accidentally triggered by another developer that's unfamiliar with the code. I'd have to think some more about how to future-proof that sort of behavior.

It'd be good to note in the comments that the queue has a precondition of expecting a pending blockchain transaction per element, and only moves forward as the transaction is confirmed.

A more robust long-term solution would be to update the DB with the ticket submission status, and have the queue check the DB for the oldest non-submitted ticket, rather than using a channel. That gives us persistence across reboots and some more options around failed transactions (where tickets in the queue would be stuck indefinitely).

Reply:
> Would checking sufficientMaxFloat prior to checking q.in address that?

I don't think so because checking `sufficientMaxFloat` requires receiving the ticket from `q.in` first (or else where does `nextTicket` come from). So, the code snippet you provided where `q.in` is drained directly into `q.redeemable` would have to look something like this:

```golang
		select {
		case maxFloat := <-q.availableMaxFloat:
			select {
			case nextTicket := <-q.in:
				if q.sufficientMaxFloat(maxFloat, nextTicket) {
					select {
					case q.redeemable <- nextTicket:
					case <-q.quit:
						return
					}
				} else {
					// Send `nextTicket` back into `q.in`
					// to add it back to the queue?
				}
			case <-q.quit:
				return
			default:
			}
		case <-q.quit:
			return
		default:
		}
```

> It'd be good to note in the comments that the queue has a precondition of expecting a pending blockchain transaction per element, and only moves forward as the transaction is confirmed.

👍 

> A more robust long-term solution would be to update the DB with the ticket submission status, and have the queue check the DB for the oldest non-submitted ticket, rather than using a channel. That gives us persistence across reboots and some more options around failed transactions (where tickets in the queue would be stuck indefinitely).

Ah yes, excellent point. We're already storing each winning ticket in the DB before we redeem it, so we're already well set up to do this. I'll create a GH issue to track this and we can follow up in a separate PR if that works for you.

Reply:
> checking sufficientMaxFloat requires receiving the ticket from q.in first (or else where does nextTicket come from)

You're right, I totally missed that.

Could we skip the sufficiency check here and rely on the check that's [within `redeemWinningTicket`](https://github.com/livepeer/go-livepeer/blob/f46945f0864e52f6ed87c20a1ac668a918a834c7/pm/recipient.go#L318-L333) ? Seems like that would also lead to re-queuing if necessary. That maybe comes at the expense of less "defense in depth" but it seems the code would be cleaner overall. If the flow is too hard to reason about right now, then whatever approach you're thinking of is fine (condvars?).

Reply:
> Could we skip the sufficiency check here and rely on the check that's within redeemWinningTicket ?

That check's purpose is a bit different. The check in `redeemWinningTicket` is to determine if we should queue a ticket when the max float is insufficient. The check here is to determine if we should pop a ticket i.e. if the most up to date max float is sufficient to cover the next ticket from the queue. We need some way to determine when a ticket should come off the queue.

> If the flow is too hard to reason about right now, then whatever approach you're thinking of is fine (condvars?).

My current preference is to revert back to using a condvar in this PR for now and then revisit this workflow separately using the DB backed approach that pulls the oldest non-confirmed winning ticket that you [mentioned](https://github.com/livepeer/go-livepeer/pull/926#discussion_r297903557) which could make a lot of this design discussion around channels vs. condvars moot.

Reply:
Rolled back the commit that replaced condvar with `q.in` so we're using a condvar again. Also, added some more clarifying comments in 1d23297